src/musl/mmap.cpp

```diff
@@ -1,124 +1,172 @@
-    Expects(false && "Mapping to file descriptor not supported");
-    errno = ENODEV;
-    return MAP_FAILED;
+  // TODO(mazunki): this is unnecessary when ::Sharing becomes a real type
+  if (util::has_flag(flags, Sharing::Private) == util::has_flag(flags, Sharing::Shared)) {
+    Expectsf(false, "sys_mmap: mapping must be either Private xor Shared");
+    return mmap_failed(EINVAL);
   }
 
-  if ((flags & MAP_ANONYMOUS) == 0) {
-    Expects(false && "We only support MAP_ANONYMOUS calls to mmap()");
-    errno = ENOTSUP;
-    return MAP_FAILED;
+  if (length == 0) {
+    Expectsf(false, "Mapping must never allocate 0 bytes");
+      return mmap_failed(EINVAL);
   }
 
-  if ((flags & MAP_FIXED) > 0) {
-    Expects(false && "MAP_FIXED not supported.");
-    errno = ENOTSUP;
-    return MAP_FAILED;
+  if (util::has_flag(Flags::Anonymous)) {
+    if (file_descriptor) {
+      Expectsf(false, "Anonymous mappings must set fd=-1, got fd={}", file_descriptor.value());  // TODO(mazunki): rename -1 when signature changes
+      return mmap_failed(EINVAL);
+    }
+    if (offset != 0) {
+      Expectsf(false, "Anonymous mappings should have offset=0, got offset={}", offset);
+      return mmap_failed(EINVAL);
+    }
   }
 
-  if (((flags & MAP_PRIVATE) > 0) && ((flags & MAP_ANONYMOUS) == 0)) {
-    Expects(false && "MAP_PRIVATE only supported for MAP_ANONYMOUS");
-    errno = ENOTSUP;
-    return MAP_FAILED;
+  // NOTE: specifying an address with non-fixed allocation is, per POSIX, only
+  // a hint. for now, we ignore this hint.
+  //
+  // this is the only reason why `uintptr_t addr` isn't const
+  if ((addr != 0) && util::missing_flag(flags, Flags::FixedOverride))  {
+    addr = 0;
   }
 
-  if (((flags & MAP_PRIVATE) > 0) && (addr != 0)) {
-    Expects(false && "MAP_PRIVATE only supported for new allocations (address=0).");
-    errno = ENOTSUP;
-    return MAP_FAILED;
-  }
+  // TODO: Implement minimal functionality to be POSIX compliant
+  // https://pubs.opengroup.org/onlinepubs/009695399/functions/mmap.html
 
-  if (((flags & MAP_SHARED) == 0) && ((flags & MAP_PRIVATE) == 0)) {
-    Expects(false && "MAP_SHARED or MAP_PRIVATE must be set.");
-    errno = ENOTSUP;
-    return MAP_FAILED;
+  if (file_descriptor) {
+    Expectsf(false, "Mapping to file descriptor is not yet implemented, got fd={}", file_descriptor.value());
+    return mmap_failed(ENOTSUP);
   }
 
-  // If we get here, the following should be true:
-  // MAP_ANONYMOUS set + MAP_SHARED or MAP_PRIVATE
-  // fd should be 0, address should be 0 for MAP_PRIVATE
-  // (address is in any case ignored)
+  if (util::missing_flag(flags, Sharing::Anonymous)) {
+    Expectsf(false, "Support for non-MAP_ANONYMOUS mappings is not yet implemented");
+    return mmap_failed(ENOTSUP);
+  }
 
-  auto* res = kalloc(length);
+  void *res = nullptr;
+
+  if (util::has_flag(flags, Flags::FixedOverride) or util::has_flag(flags, Flags::FixedFriendly)) {
```

Comment:
Wait - this is not just reorganizing stuff it it? Is this copy-pasted in from somewhere else or is it new functionality?

Reply:
Yep, these changes will come later.